feat: make it possible to check for outdated components in JSON flows

ogabrielluiz opened the pull request:
Adds the check command which helps users check if their flows have outdated components and update them if they so choose.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added a CLI command to check flows for outdated components, show diffs/breaking changes, and optionally update components (auto or interactive).
  - Run command now supports a --check option to prevent execution when outdated components are detected.
  - CLI displays help when no arguments are provided.

- Chores
  - Simplified all starter project templates by removing dependency metadata from components, reducing noise and file size.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ogabrielluiz committed:
refactor: Comment out dependency analysis in build_component_metadata function

- Removed the call to analyze_component_dependencies from the build_component_metadata function to improve load times.
- Added a TODO comment indicating the need to relocate this functionality to a more appropriate location in the codebase.

These changes enhance performance while maintaining the existing functionality for future implementation.

ogabrielluiz committed:
feat: Add CLI check command for outdated components in flows

- Introduced a new CLI command to detect and update outdated components in JSON flows.
- Implemented functionality to load specific components based on flow metadata, check for outdated components using code hash comparison, and handle updates interactively or automatically.
- Enhanced logging for better debugging and user feedback during the update process.

This addition improves the usability of the framework by allowing users to maintain up-to-date components in their flows efficiently.

ogabrielluiz committed:
feat: Implement component check before execution in CLI

- Added an asynchronous function to check for outdated components in JSON flows before execution.
- Introduced a new CLI option to enable this check, enhancing user experience by ensuring components are up to date.
- Improved error handling to provide detailed feedback on outdated components found during the check.

This addition promotes better maintenance of flow components and prevents potential runtime issues.

ogabrielluiz committed:
feat: Add check command to Langflow CLI for outdated components

- Introduced a new command in the Langflow CLI to check for outdated components in flows, enhancing user experience by ensuring components are current.
- Updated the CLI structure to include the check command alongside existing serve and run commands.
- Improved documentation for the new command to facilitate user understanding and usage.

This addition promotes better maintenance of flow components and aligns with recent enhancements for component checks.

ogabrielluiz committed:
feat: add check command to lfx CLI

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ogabrielluiz committed:
feat: Add unit tests for check command functionality

- Introduced comprehensive unit tests for the check command, covering various scenarios including component detection, outdated checks, and error handling.
- Implemented tests for both synchronous and asynchronous functions, ensuring robust validation of the check command's behavior.
- Enhanced test coverage for component mapping and breaking change detection, contributing to improved reliability and maintainability of the codebase.
- Included tests for integration with real flow files, ensuring the command operates correctly in practical scenarios.

These additions strengthen the testing framework and promote confidence in the check command's functionality.

ogabrielluiz committed:
Merge branch 'main' into add-check-command

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ogabrielluiz committed:
fix: Update code diff generation and improve component update handling

- Adjusted the `generate_code_diff` function to skip the correct headers in the diff output, enhancing accuracy in code comparison.
- Replaced shallow copy with deep copy for component updates in `apply_component_update`, ensuring that nested structures are correctly preserved during updates. This change improves data integrity and prevents unintended mutations.

ogabrielluiz committed:
fix: Ensure minimal dependency information is set in component metadata

- Updated the `build_component_metadata` function to always initialize the `dependencies` field in the frontend node's metadata with default values, enhancing consistency and preventing potential issues when analysis is disabled.

ogabrielluiz committed:
test: Mark dependency analysis tests as expected failures

- Added `@pytest.mark.xfail` to tests for building component metadata with external, optional, and real dependencies, indicating that the dependency analyzer is currently commented out and returns minimal stub data. This change helps clarify the expected behavior during testing and maintains test integrity until the analyzer is fully implemented.

ogabrielluiz committed:
fix: Enhance component loading with security checks and error handling

- Added security checks to ensure only trusted component namespaces are loaded, preventing potential vulnerabilities.
- Implemented error handling for component instantiation failures, logging warnings for invalid paths and instantiation issues to improve robustness and traceability during component loading.

ogabrielluiz committed:
test: Add comprehensive security tests for load_specific_components function

- Introduced a new test suite for the `load_specific_components` function to validate security measures against various attack vectors, including blocking malicious module imports and handling invalid paths.
- Ensured that both valid and invalid component paths are processed correctly, maintaining robust security validation.
- Implemented tests for case sensitivity and strict prefix validation to prevent potential bypass attempts.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

mpawlow commented on src/backend/base/langflow/__main__.py:
```diff
@@ -45,12 +45,14 @@
 
 # Add LFX commands as a sub-app
 try:
+    from lfx.cli.check import check_command_sync
```
### () [Question] Is there a reason not to import (an augmented) `check_command` function directly rather than the sync wrapper?

- Noticed that `run` uses an annotation to wrap the async function as sync
- Wondering why the same approach wasn't followed?

```python
from functools import partial
from asyncer import syncify

@partial(syncify, raise_sync_error=False)
async def run(
...
```

mpawlow commented on src/lfx/src/lfx/cli/check.py:
```diff
@@ -0,0 +1,774 @@
+        return True
+
+    # Check for output type changes
+    for current_out in current_outputs:
```
### () [Minor] Performance: O(n^2)

- A loop within a loop could be optimized by pre-indexing current_outputs / latest_outputs into a Sets
- However, this might be overkill given that the number of outputs for current / latest is likely always < 10

mpawlow commented on src/lfx/src/lfx/cli/check.py:
```diff
@@ -0,0 +1,774 @@
+
+    if current_code != latest_code:
+        diff_result = generate_code_diff(current_code, latest_code)
+        changes["code_diff"] = diff_result
```
### () [Minor] Calculating the code differences could potentially be an expensive operation

- Not sure how much code a component could potentially contribute
- Wondering if it makes sense to not calculate code differences by default and specify a new CLI command-line option that toggles on calculating code differences
- Please ignore if calculating code differences is of critical importance to this check operation

mpawlow commented on src/lfx/src/lfx/cli/check.py:
```diff
@@ -0,0 +1,774 @@
+"""CLI check command for detecting and updating outdated components in flows."""
```
### () [Question] What functions should be exposed for the default import for `lfx.cli` in `__init__.py`?

- Right now it's only the `server_command`

```python
from lfx.cli.commands import serve_command

__all__ = ["serve_command"]
```

- In the tests, it's checking for the `check_command_sync`, `check_command` and `check_components_before_run` functions

```python
    def test_check_command_functions_exist(self):
        """Test that check command functions are available."""
        from lfx.cli.check import check_command, check_command_sync

        assert callable(check_command_sync)
        assert callable(check_command)

    def test_run_command_check_function_exists(self):
        """Test that run command check function exists."""
        from lfx.cli.run import check_components_before_run

        assert callable(check_components_before_run)
```

jordanrfrazier commented on src/lfx/src/lfx/cli/check.py:
```diff
@@ -0,0 +1,774 @@
+            console.print(f"  • {safe_updates} safe update(s)")
+        if breaking_updates > 0:
+            console.print(f"  • {breaking_updates} breaking update(s)")
+            console.print("[yellow]⚠️  Please test your flow thoroughly due to breaking changes[/yellow]")
```
I think the `--upgrade` flag is doing breaking updates without the `--force` being specified - see https://gist.github.com/jordanrfrazier/acb9ecddbc0c4a1a4888aafabfad1f21

mpawlow reviewed (changes requested):
### Code Review 1 + Feedback

- My apologies, I had this (partial) code review in flight before the last initiatives pivot
- Please feel free to ignore the comments tagged with "Minor" severity (optional)

ogabrielluiz committed:
feat: Add in-place update option for flow components

- Introduced an `--in-place` option to the `check_flow_components` and `check_command` functions, allowing users to update the input file directly.
- Enhanced error handling to ensure that either an output file or in-place update is specified when applying updates.
- Updated documentation and help messages to reflect the new functionality, improving user guidance for component updates.

ogabrielluiz committed:
test: Enhance unit tests for in-place update functionality

- Refactored the `sample_flow_data` fixture for improved clarity and consistency.
- Added comprehensive tests for the `--in-place` option in `check_flow_components`, ensuring correct behavior with and without the `--update` flag.
- Implemented tests to verify that the original file structure is preserved during in-place updates and that error handling is robust when required parameters are missing.
- Included scenarios for interactive mode to confirm user prompts are handled correctly during updates.

ogabrielluiz committed:
refactor: Simplify command imports and remove synchronous check command

- Consolidated imports for `check_command` and `serve_command` in the LFX CLI.
- Removed the synchronous `check_command_sync` function, transitioning to an asynchronous implementation for improved performance and consistency.
- Updated related tests to reflect the changes in command availability.

ogabrielluiz committed:
fix: Add error handling for unavailable settings service in check_flow_components

- Implemented a check to return an error message if the settings service is not available, preventing potential failures when loading component types.
- Enhanced robustness of the `check_flow_components` function by ensuring it handles the absence of necessary services gracefully.

ogabrielluiz committed:
refactor: Improve component loading logic in check_flow_components

- Enhanced the component loading process by implementing a fallback mechanism to load all component types if selective loading returns an empty dictionary.
- Added debug logging to provide clearer insights into the loading process, improving traceability and debugging capabilities.

ogabrielluiz committed:
refactor: Optimize output type checking in breaking changes validation

- Improved the efficiency of the `check_breaking_changes` function by pre-indexing latest outputs by name, allowing for O(1) lookup during type comparisons.
- Simplified the logic for checking output type changes, enhancing readability and performance.

ogabrielluiz commented on src/lfx/src/lfx/cli/check.py:
```diff
@@ -0,0 +1,774 @@
+                await apply_component_update(flow_data, component, all_types_dict)
+                applied_updates += 1
+    else:
+        # Interactive mode - prompt for each component individually
```
Fixed.

ogabrielluiz committed:
feat: Add show_diff option to check_flow_components for detailed code changes

- Introduced a `--show-diff` option in the `check_command` and `check_flow_components` functions, allowing users to view code differences when checking for outdated components.
- Updated the `check_component_outdated` and `analyze_component_changes` functions to conditionally calculate and display code diffs based on the new option, enhancing user feedback during updates.
- Improved interactive prompts to show all diff blocks, providing clearer insights into changes made to components.

ogabrielluiz committed:
test: Add unit tests for interactive mode and show_diff functionality in check_flow_components

- Introduced tests to validate that interactive mode fails early when neither --output nor --in-place is specified, ensuring proper error handling.
- Added tests for the --show-diff option to confirm that code differences are calculated correctly based on user input, including scenarios for both true and false values.
- Enhanced existing test cases to ensure robust behavior when no code changes are present, maintaining stability in the check_flow_components function.

ogabrielluiz committed:
feat: Enhance interactive prompts and output handling in check_flow_components

- Added functionality to allow interactive prompts to use stderr when output is set to stdout, preventing JSON output from mixing with prompt messages.
- Updated the `handle_updates` function to handle output paths more robustly, including writing to stdout or specified files.
- Improved the `prompt_for_component_update` function to accept an output parameter, ensuring consistent behavior across different output scenarios.
- Enhanced error messages to clarify output requirements when applying updates, improving user experience during component updates.

ogabrielluiz committed:
test: Add unit tests for stdout output handling in check_flow_components

- Introduced tests to validate the behavior of the --output - option, ensuring it correctly writes to stdout.
- Added tests to confirm that stdout takes precedence over in-place updates, preserving original file content.
- Implemented tests for interactive mode to verify that --output - is accepted without errors, enhancing overall test coverage for output scenarios.

ogabrielluiz committed:
feat: Add backup option for in-place updates in check_flow_components

- Introduced a `--backup` option to create a backup file before modifying the original flow file in place, enhancing data safety.
- Implemented logic to validate that the backup option requires the in-place flag, ensuring correct usage.
- Updated the `handle_updates` function to handle backup creation and logging, providing feedback on backup status to the user.
- Enhanced the `check_command` function to include the new backup option in the command-line interface.

ogabrielluiz committed:
test: Add unit tests for backup option in check_flow_components

- Introduced tests to validate the behavior of the --backup option during in-place updates, ensuring backups are created and contain original content.
- Added tests to confirm that backups are not created when --no-backup is specified and that the backup option only works with in-place updates.
- Enhanced test coverage for interactive mode to verify backup functionality when user accepts updates, improving overall reliability of the backup feature.

ogabrielluiz committed:
feat: Enhance error handling in check_flow_components

- Added a mechanism to capture and report errors during component checks, improving feedback for components lacking code.
- Updated the return structure to include a list of check errors and their counts, allowing users to identify issues more easily.
- Enhanced console output to display any check errors encountered, providing clearer insights into the component validation process.

ogabrielluiz committed:
test: Add unit test for handling missing code in component checks

- Introduced a test to verify that the `check_component_outdated` function correctly identifies and reports an error when both the node and component lack code.
- Ensured that the error message includes relevant details, improving feedback for users when components are missing necessary code attributes.

ogabrielluiz committed:
Merge branch 'main' into add-check-command

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 2/3)

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 3/3)

ogabrielluiz committed:
Merge branch 'main' into add-check-command

# Conflicts:
#	src/lfx/src/lfx/_assets/component_index.json

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 2/3)

ogabrielluiz committed:
refactor: remove redundant backup validation in check_command function

jordanrfrazier reviewed (approved):
Code looks good to me. if it's tested and working manually then lgtm

ogabrielluiz committed:
refactor: enhance update handling and error logging in handle_updates function

ogabrielluiz committed:
refactor: improve error handling in check_components_before_run function

ogabrielluiz committed:
test: add unit tests for apply_component_update and check_components_before_run functions
